beam/tools/src/main/java/cz/o2/proxima/beam/tools/groovy/BeamStream.java

```diff
@@ -1497,69 +1510,145 @@ static int getPort(int preferredPort, int seed) {
           : ((ThreadLocalRandom.current().nextInt() ^ seed) & Integer.MAX_VALUE) % 50000 + 10000;
     }
 
-    private class Server extends NanoHTTPD {
+    private class CollectService extends CollectServiceGrpc.CollectServiceImplBase {
 
-      Server(int port) {
-        super(port);
-      }
+      private final List<CompletableFuture<Void>> unterminatedCalls =
+          Collections.synchronizedList(new ArrayList<>());
 
       @Override
-      public NanoHTTPD.Response serve(NanoHTTPD.IHTTPSession session) {
-        synchronized (RemoteConsumer.this) {
-          try {
-            consumer.accept(deserialize(session.getInputStream()));
-            return newFixedLengthResponse("OK");
-          } catch (IOException ex) {
-            throw new RuntimeException(ex);
+      public StreamObserver<Item> collect(StreamObserver<Response> responseObserver) {
+        CompletableFuture<Void> terminateFuture = new CompletableFuture<>();
+        unterminatedCalls.add(terminateFuture);
+        return new StreamObserver<Item>() {
+
+          @Override
+          public void onNext(Item item) {
+            consumer.accept(
+                ExceptionUtils.uncheckedFactory(
+                    () -> deserialize(item.getSerialized().newInput())));
           }
-        }
+
+          @Override
+          public void onError(Throwable throwable) {
+            log.warn("Error in collect()", throwable);
+            terminateFuture.completeExceptionally(throwable);
+          }
+
+          @Override
+          public void onCompleted() {
+            responseObserver.onNext(
+                Response.newBuilder().setStatus(200).setStatusCode("OK").build());
+            responseObserver.onCompleted();
+            terminateFuture.complete(null);
+          }
+        };
+      }
+
+      public void awaitAllClosed() {
+        ExceptionUtils.ignoringInterrupted(
+            () ->
+                CompletableFuture.allOf(unterminatedCalls.toArray(new CompletableFuture[] {}))
+                    .get());
+      }
+    }
+
+    private class Server implements AutoCloseable {
+
+      private final int port;
+      private final io.grpc.Server server;
+      private final CollectService service = new CollectService();
+
+      private Server(int port) {
+        this.port = port;
+        server =
+            ServerBuilder.forPort(port)
+                .executor(Executors.newCachedThreadPool())
```

Comment:
this lead to memory leaks in the past, fixed ForkJoinPool is recommended for the best performance and came out best from our measurements.